src/svd/util.py

```diff
@@ -0,0 +1,386 @@
+        return True
+
+    def is_reg_selected(self, reg: svd.Register) -> bool:
+        if self.address_range is not None:
+            reg_addr_range = reg.address_range
+            if not _ranges_overlap_inclusive(
+                reg_addr_range.start, reg_addr_range.stop - 1, *self.address_range
+            ):
+                return False
+
+        if (
+            self.content_status == BuildSelector.ContentStatus.WRITTEN
+            and not reg.written
+        ):
+            return False
+
+        if (
+            self.content_status == BuildSelector.ContentStatus.MODIFIED
+            and not reg.modified
+        ):
+            return False
+
+        return True
+
+    def is_field_selected(self, field: svd.Field) -> bool:
+        # svada does not have a way to check written status at this level currently
+        if (
+            self.content_status == BuildSelector.ContentStatus.MODIFIED
+            and not field.modified
+        ):
+            return False
+
+        return True
+
+
+class DeviceBuilder:
+    """Used to populate peripheral registers and output the register contents in various formats."""
+
+    def __init__(
+        self, device: svd.Device, enforce_svd_constraints: bool = True
+    ) -> None:
+        self._device = device
+        self._enforce_svd_constraints = enforce_svd_constraints
+        self._written_peripherals = {}
+        self._cached_device_periph_map = []
+        self._cached_periph_reg_maps = {}
+
+    @property
+    def device(self) -> svd.Device:
+        """The device structure the builder was initialized with."""
+        return self._device
+
+    @property
+    def written_peripherals(self) -> list[svd.Peripheral]:
+        """The list of peripherals that have been written to as part of the API calls."""
+        return list(self._written_peripherals.values())
+
+    def build_bytes(self, selector: BuildSelector = BuildSelector()) -> bytearray:
+        """Encode device content as bytes.
+
+        :param selector: selected parts of the device.
+        :returns: content bytes.
+        """
+        memory = self.build_memory(selector)
+        out = bytearray()
+
+        for (addr_a, value_a), (addr_b, value_b) in pairwise(memory.items()):
+            if not out:
+                out.append(value_a)
+
+            num_empty = addr_b - addr_a - 1
+            if num_empty > 0:
+                # TODO: 0 may not always be valid
+                out.extend([0] * num_empty)
+
+            out.append(value_b)
+
+        return out
+
+    def build_memory(self, selector: BuildSelector = BuildSelector()) -> dict[int, int]:
+        """Encode device content as a mapping between address and value.
+
+        :param selector: selected parts of the device.
+        :returns: content memory map.
+        """
+        memory = {}
+
+        for peripheral in self._device.values():
+            if not selector.is_periph_selected(peripheral):
+                continue
```

Comment:
Non-blocking: it would be nice to print a warning if no peripheral gets selected, and/or if the selector references peripheral names that don't exist in the SVD.

Reply:
Added checks for these